Yatima/Tests/ExprGenerator.lean

```diff
@@ -0,0 +1,349 @@
+def randCid (g : gen) : Cid × gen :=
+  let multiHash : Multihash := ⟨0, 0, .mk #[0xC0DE]⟩
+  let (ver, g) := randNat g 0 1
+  let (num, g) := randNat g 0 1
+  (⟨ver, num, multiHash⟩, g)
+
+def randSort (g : gen) : Yatima.Expr × gen :=
+  let (anonCid, g) := randCid g
+  let (metaCid, g) := randCid g
+  let univCid := ⟨⟨anonCid⟩, ⟨metaCid⟩⟩
+  (.sort univCid, g)
+
+def randConst (g : gen) : Yatima.Expr × gen :=
+  let (name, g) := randName g
+  let (anonCid, g) := randCid g
+  let (metaCid, g) := randCid g
+  let constCid := ⟨.mk anonCid, .mk metaCid⟩
+  (.const name constCid [], g)
+
+def randLit (g : gen) : Yatima.Expr × gen :=
+  let (isNat, g) := randBool g
+  if isNat then 
+    let (num, g) := randNat g 0 100
+    (.lit $ .nat num, g) else
+    let (str, g) := randName g
+    (.lit $ .str str, g)
+
+def randVar (depth : Option Nat) (g : gen) (isBdd : Bool := false): Yatima.Expr × gen :=
+  let (name, g) := randName g
+  match depth with
+    | none => 
+      let (idx, g) := randNat g 0 100
+      (.var name idx, g)
+    | some dep => 
+      let (isBddR, g) := randBool g
+      if isBdd || isBddR then
+        let (idx, g) := randNat g 0 dep
+        (.var name idx, g) else
+        let (idx, g) := randNat g (dep + 1) 100
+        (.var name idx, g)
+
+partial def assembleExprAux (head : Hole) (g : gen) : ExprGen Yatima.Expr := do
+  let hExpr ← getValue? head
+  match hExpr with
+    | none => unreachable!
+    | some expr => match expr with
+      | .app funcHole inputHole           => 
+        let (g₁, g₂) := RandomGen.split g
+        return .app (← assembleExprAux funcHole g₁) (← assembleExprAux inputHole g₂)
+      | .lam typeHole bodyHole            =>
+        let (name, g) := randName g
+        let (g₁, g₂) := RandomGen.split g 
+        return .lam name default (← assembleExprAux typeHole g₁) (← assembleExprAux bodyHole g₂)
+      | .pi typeHole bodyHole             =>
+        let (name, g) := randName g 
+        let (g₁, g₂) := RandomGen.split g
+        return .pi name default (← assembleExprAux typeHole g₁) (← assembleExprAux bodyHole g₂)
+      | .letE typeHole valueHole bodyHole =>
+        let (name, g) := randName g
+        let (g₁, g₂) := RandomGen.split g
+        let (g₂, g₃) := RandomGen.split g₂
+        return .letE name (← assembleExprAux typeHole g₁) 
+                          (← assembleExprAux valueHole g₂) 
+                          (← assembleExprAux bodyHole g₃)
+      | .fix bodyHole                     =>
+        let (name, g) := randName g
+        return .fix name (← assembleExprAux bodyHole g) 
+      | .leaf leafType => match leafType with
+        | .sort        => return randSort g  |>.1
+        | .const       => return randConst g |>.1
+        | .lit         => return randLit g   |>.1
+        | .var varType => 
+          let depth? ← getBinderDepth head
+          return randVar depth? g |>.1
+
+partial def run (g : gen) : ExprGen Yatima.Expr := do
```

Comment:
@DanielRrr the other functions are partial because Lean is unable to automatically prove termination for them due to their use of recursion.

But @mpenciak this one doesn't need to be partial. `partial` isn't viral: we can think of partial functions as constants
```suggestion
def run (g : gen) : ExprGen Yatima.Expr := do
```